### ISSUE TITLE
Sparse matrix integer support

### ISSUE THREAD

**Issue by dloney** [cupy/cupy#3650]
The current Cupy sparse matrix initialization does not support integer matrices as in the case of Scipy. Was integer support left out intentionally from the sparse matrix routines?

**Comment by leofang** [comment]
@anaruse Am I right that cuSPARSE used to support only float, double, float complex, and double complex, but now with the new Generic API it also supports char/int8_t and int/int32_t?

https://docs.nvidia.com/cuda/cusparse/index.html#cusparse-generic-enum-data-types

**Comment by anaruse** [comment]
Yes, cuSPARSE generic API also supports the integer data types. However, note that the combinations of supported data types vary from function to function.
cusparseSpMV(): https://docs.nvidia.com/cuda/cusparse/index.html#cusparse-generic-function-spmv

**Comment by dloney** [comment]
It sounds like it would be the first step would be to allow support for additionally types during matrix construction. Would it also be good to include a type check within each function to warn about unexpected behavior if the matrices are different formats?

**Comment by leofang** [comment]
@dloney I wonder if you have done some tests using cuSPARSE's API directly for its int8/int32 supports and compare the performance with, say, float32? The limited support in cuSPARSE seems nontrivial and so worrying to me, because we'd need to set up a dtype guard to every single function. I wanna know if it's worth the effort. 

Also, what would be the use cases you have in mind for integer sparse matrices? I understand having better compatibility with SciPy is always a good thing, but not if the underlying library is lacking some critical pieces...I imagine if we don't set up the guard right, some users would come and complain "why is there a type conversion happening at this line?" And we'd have countless branches to be covered in our tests.

Just my two cents 😅 The PFN team may have already looked into this, so you can definitely wait for their feedbacks!
